오픈 마켓 [STEP 2] Chacha, Jiseong

yim2627 opened the pull request:
# Step 2 PR

안녕하세요 도미닉! @AppleCEO  👍

지성, 차차입니다.

Step2 구현이 완료되어 PR 제출 드립니다.

이번 스텝도 잘 부탁드립니다!

감사합니다😁

## 고민했던 점 

### 1. Modern CollectionView 구현

기존의 방식과 달리 애플에서 새롭게 제시하는 modern collectionView 구현 방식을 사용하여 프로젝트를 진행했습니다. 
이 때 compositional Layout과 DiffableDataSource를 주로 활용하였으며, 이 들의 장점을 체감했던 것 같습니다.

- Compositional Layout
    - 입맛에 맞게 유연하고 빠르게 구성할 수 있다.
    - section/group/item을 쌓는 모습이 시각적으로 나타난다

- DiffableDataSource
    - 데이터를 관리하고 collectionView에 대한 cell을 제공하는데 사용된다.
    - collectionView의 데이터와 UI를 간편하고, 효율적이게 관리할 수 있다.

이러한 내용들에 기반해서 이번 프로젝트는 modern CollectionView 구현 방식을 사용했습니다!

### 2. List/Grid collectionView 스위치하기

두 개의 collectionview에 각각의 Cell을 추가하여 사용할 지, 하나의 collectionView에 각각의 cell을 상황에 따라 바꿔 사용할 지 고민해보았습니다.

- 두 개의 collectionview를 사용하는 경우
    - `isHidden`으로만 쉽게 switching을 관리할 수 있다.
- 하나의 collectionview에 cell이 두 개 바뀌는 경우
    - 매번 switch될 때마다 새로 collectionView의 layout, dataSource 등을 설정해줘야 한다. 

고민해본 결과, SegmentedControl을 사용하여 하나의 collectionView에 List/ Grid로 레이아웃을 변경할 때마다 새로이 layout을 설정하는 것의 비용이 더 클 것이라고 생각하여 ListLayoutCollectView와 GridLayoutCollectionView를 각각 만들어 레이아웃에 만든 셀을 추가시키고, 콜렉션뷰의 인스턴스를 한번씩만 생성하도록 하는 방식을 채택하였습니다.

첫 화면이 ListLayoutCollectionView 이므로 앱이 시작되고 데이터가 불러와졌을 때 해당 뷰의 인스턴스를 생성하였고 화면이 스위칭되었을 때 GridLayoutCollectionView의 인스턴스를 생성해주었으며, 각 뷰가 스위칭 될 때마다 현재 보이는 뷰는 hidden되지않고 보이지 않는 뷰는 hidden되게 만들어 뷰를 올렸습니다. alpha값을 변경하여 뷰를 숨기는 것과 고민했으나, alpha 값을 0으로 주게 되면 보기에는 사라지지만 뷰가 실제로는 그려지기에 메모리 관점에서 `isHidden`을 사용하는 것이 뷰를 그리지 않는 방법이라고 하여 비용 관점에서 효율적이라고 생각했습니다.


### 3. 스토리보드 없이 코드로 하게된 이유

![](https://minjunkweon.github.io/assets/images/image-20190814150748264.png)


- 리팩터링이 더 편하다고 생각하였습니다. 
    - 변경사항을 편하게 추적하고 괸리할 수 있을 것 같았습니다. 
        - 협업의 경우 보통 git을 사용하여 협업을 하는데 이전 커밋과의 diff를 살펴볼 때 스토리보드의 경우는 XML로 올라가 있기때문에 비교가 직관적이지 못하다고 생각하였습니다.
    - 레이아웃이 꼬인 경우, 스토리보드에선 어느 부분부터 꼬인 것인지 직관적으로 보이지 않고, 레이아웃이 잘못됐다는 오류 또한 Xcode의 오류때문인지 정상적인 것인데 잘못됐다고 나오거나 오류가 나오다 마는 경우도 있었습니다.
- 코드로 UI를 작성하는 것을 경험해보고 싶었습니다.
    - modern Cell, modern collectionView의 경우 코드로 Configuration을 지정하여 주기 때문입니다.
- 간단한 레이아웃을 설정하는 부분이라면 재사용을 고려할 수 있었습니다.
- 위 사진처럼 코드로 짜는 UI가 스토리보드로 짜는 UI의 범주를 포함하고있어 스토리보드로만 UI를 구성하는 것은 한계가 있다고 생각하였습니다.


## 궁금한 부분 

### 1. 상수 관리에 관한 부분 (autolayout의 constant)

보통 상수를 관리하기 위해 namespace를 사용하는 것으로 알고 있습니다. 프로젝트를 진행하면서 layout constant 값을 임의로 주게 되었는데 이 또한 namespace로 관리해야할지 궁금하여 질문드립니다. 모든 제약값을 namespace에 넣게 되면 코드가 더 길어지고 복잡해지지 않을까하는 생각에 결국에는 직접 적어주게 되었는데 어떤 것이 올바른 선택인지 궁금합니다!

### 2. 보통 safe Area 넘어 view의 상단까지 레이아웃을 주는지 궁금합니다. 

UINavigationController를 사용하게 되면 상단에 navigation bar가 생기게 되는데, 이 뒷 부분까지 view가 넘어오도록 layout을 주어야 할지, 아니면 safeAreaLayoutGuide의 topAnchor까지만 주어야 하는지 궁금합니다.
(보여지는 차이는 그렇게 크지 않은 것 같은데, safeArea를 넘어서까지 layout을 주게되니 종종 navigationbar와 statusbar의 배경이 사라지는(?)문제가 있었어서 결국에는 safeArea의 topAnchor에 맞추게 되었습니다.)

사람들이 많이 사용하는 어플리케이션들을 살펴보아도 navigationBar나, statusBar에 색상을 지정하여 뷰가 safeArea 넘어서까지 올라감에도 불구하고 보이지 않게 만든 것인지, 아니면 safeArea top에 view의 top을 맞춘 것인지 구별을 할 수 없어 현업에서는 이를 어떻게 설정하는지 궁금합니다!

### 3. defaultContentConfiguration()을 사용했을 때 Image의 크기?

default 스타일의 cell을 사용하게 되는 경우 기본적으로 제공받는 image의 크기를 어떻게 보통 조절하는지 궁금합니다. 
찾아본 바로는 imageLayoutGuide를 사용하여 layout을 잡거나, 아래와 같이 maximumSize 혹은 reservedSize를 쓰는 것으로 알고 있는데, 이미지 자체를 resizing하기도 하는지 주로 어떤 방식을 사용하는지 궁금합니다! (저희는 최대 크기만 제한하여 이미지 크기를 통일해주었습니다)

`content.imageProperties.maximumSize = CGSize(width: 50, height: 50)`

### 4. Deployment Target

iOS14부터 사용할 수 있는 modern한 것들을 사용하였는데, 간단히 앱스토어에서 찾아본 결과 잘 알려진 어플리케이션의 경우 버젼 범주가 더 넓었습니다.

여기서 든 의문은 회사가 modern한 코드를 사용하는 경우, iOS14 이상을 지원하는 코드와 그 이하 버젼을 지원하는 코드를 따로 만들어 앱을 구현하는지가 궁금합니다.

ChaminLee committed:
feat: 스토리보드 삭제 및 SceneDelegate window 생성 코드 구현 #6

yim2627 committed:
feat: LayoutSegmentedControl 타입 생성 #6

- ViewController 내 ConfigUI 메서드 구현

ChaminLee committed:
feat: productCollectionView 생성 및 ProductListLayoutCell 구현 #6

- 모델에 Hashable 타입 채택
- productListLayoutCell layout 코드 구현
- NSMutableAttributedString extension하여 커스텀 텍스트 구현
- productListLayoutCell은 defaultContentConfiguration 방식을 사용하여 초기화
- ViewController내 createListLayout 메서드 구현

yim2627 committed:
feat: productCollectionView Layout, DataSource 구현 #6

- fetchProductData, UI 업데이트 비동기 메서드 구현
- Cell의 StockLabel 색상 변경 및 Cell 클래스 내 오타 수정
- NSMakeRange -> NSRange 변경
- 기존 defaultContentConfiguration을 defaultConfiguration 메서드로 변경
- ViewController내 ListCollectionView 관련 메서드 Extension, Mark로 분리

ChaminLee committed:
refactor: ProductListLayoutCell layout 수정 및 ImageLoader 구현 #6

- stockLabel layout 변경
- createPriceText내 attributedText 생성 로직 변경
- defaultContent내 layout 조정 및 label 속성값 변경
- NSMutableAttributedString의 strikeThroughStyle 로직 수정

yim2627 committed:
feat: ProductGridLayoutCell 구현 #6

- AttributedTextCreator 타입 구현
- NSMutableAttributedString extension 구현
- ProductListLayoutCell내 로직 변
- ProductGridLayoutCell 내부 컴포넌트 및 레이아웃 구현

ChaminLee committed:
feat: GridCollectionView 구현 #6

- ProductGridLayoutCell내 레이아웃 및 속성 변경
- GridCollectionView 레이아웃 적용 및 셀 등록

yim2627 committed:
feat: SegmentedControl 버튼 액션 구현 #6

- CollectionView 전환 구현

ChaminLee committed:
feat: Double에 decimal 적용하는 extension 구현 #6

- ProductListLayoutCell 레이아웃 조정
- Currency내 unit 연산 프로퍼티 추가

yim2627 committed:
feat: ActivityIndicatorView 구현 #6

- segment -> layoutSegmentedControl 네이밍 변경
- 접근제어자 추가

ChaminLee committed:
feat: 삼품 등록 버튼 구현 #6

- ViewController -> MainViewController 네이밍 변경
- collectionView layout 적용 코드 구현
- 상품 등록 버튼 터치시 modal을 띄워주도록 구현

yim2627 committed:
refactor: Layout 중복 제거 #6

ChaminLee committed:
refactor: layout 및 크기 조정 #6

- segmentedControl 크기 변경
- MainViewController내 사이즈 변경
- GridCell 내 이미지 사이즈 변경

AppleCEO commented on OpenMarket/OpenMarket/Cells/ProductGridLayoutCell.swift:
```diff
@@ -0,0 +1,113 @@
+        super.init(coder: coder)
+    }
+    
+    private let productStackView: UIStackView = {
```
스텍뷰의 이름이 명확하네요! 👍

AppleCEO commented on OpenMarket/OpenMarket/Cells/ProductGridLayoutCell.swift:
```diff
@@ -0,0 +1,113 @@
+        
+        productNameLabel.text = product.name
+        productImageView.image = ImageLoader.loadImage(from: product.thumbnail)
+        
```
여기 한 줄을 띄운 이유가 있을까요?
코드에는 공백이나 줄바꿈 하나에도 의미가 있어야 합니다.
의도를 나타낼 수 있는 최소한의 줄바꿈을 추천드려요 (무조건 적은 줄의 코드가 좋다는 뜻은 아님)

ChaminLee replied:
의미없는 공백은 지워두도록 하겠습니다!

AppleCEO commented on OpenMarket/OpenMarket/Cells/ProductGridLayoutCell.swift:
```diff
@@ -0,0 +1,113 @@
+            productStackView.addArrangedSubview($0)
+        }
+        
+        productPriceStackView.setContentCompressionResistancePriority(.defaultHigh, for: .vertical)
+        
+        self.contentView.addSubview(productStackView)
```
self 를 붙인 뷰와 안 붙인 뷰는 어떤 차이가 있을까요?

yim2627 replied:
현재 상황에서는 차이가 없을 것 같습니다. 
저희가 알고 있는 일반적인 상황에서의 self는 인스턴스 프로퍼티/메서드임을 명시하기위해 사용한다라고 알고있습니다.
예를들어 인스턴스 프로퍼티와 지역변수의 이름이 같아 구분이 모호할 때 self를 사용해준다고 이해하고 있습니다.

AppleCEO commented on OpenMarket/OpenMarket/Cells/ProductListLayoutCell.swift:
```diff
@@ -0,0 +1,93 @@
+}
+
+class ProductListLayoutCell: UICollectionViewListCell {
+    private var productItem: ProductDetail?
+    private lazy var listContentView = UIListContentView(configuration: defaultConfiguration())
+    private var stockLabel = UILabel()
+    private var stockLabelLayouts: [NSLayoutConstraint]?
```
접근자를 설정해 부분 좋습니다! 💯

AppleCEO commented on OpenMarket/OpenMarket/Cells/ProductListLayoutCell.swift:
```diff
@@ -0,0 +1,93 @@
+    }
+    
+    func updateWithProduct(from newProductItem: ProductDetail) {
+        guard productItem != newProductItem else {
+            return
+        }
```
여기 if문을 사용하면 어떨까요?
이렇게 작성하면 부정(!=)의 부정(else) 라서 코드를 이해하기 더 어려워집니다.

ChaminLee replied:
동일한 의미를 갖지만, 말씀주신 것 처럼 아래 코드처럼 더 명확하게 변경하겠습니다

```swift
if productItem == newProductItem {
    return
}
```

AppleCEO commented on OpenMarket/OpenMarket/Cells/ProductListLayoutCell.swift:
```diff
@@ -0,0 +1,93 @@
+            return
+        }
+        
+        [listContentView, stockLabel].forEach {
+            contentView.addSubview($0)
+            $0.translatesAutoresizingMaskIntoConstraints = false
+        }
```
아직은 축약형을 사용하지 않는 것을 더 추천합니다.
영어처럼 읽히는 코드가 좋은 코드라고 생각합니다.

```swift
for view in [listContentView, stockLabel] {
  contentView.addSubView(view)
  view.translatesAutoresizingMaskIntoConstatraints = false
} 
```

작성하신 코드와 어떤 것이 더 쉽게 읽히나요?

yim2627 replied:
forEach를 사용하고, 축약형을 사용하지 않는 방식으로 변경하겠습니다! 😁

```swift
[listContentView, stockLabel].forEach { view in
     contentView.addSubview(view)
     view.translatesAutoresizingMaskIntoConstraints = false
}
```

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,183 @@
+import UIKit
+
+private enum Section: Hashable {
```
Section 이라는 이름도 너무 짧은 것 같네요.
이러한 이름은 예약어로 등록되어 있어서 나중에 문제가 생길 수 있습니다.
예약어가 아니더라도 어떤 섹션인지 불분명해서 조금 더 길고 명확한 이름이 좋을 것 같습니다~

ChaminLee replied:
`Section` -> `ProductSection`으로 명확하게 변경해주려고 합니다!

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,183 @@
+    private var productData: [ProductDetail] = []
+    
+    private lazy var activityIndicator: UIActivityIndicatorView = {
+        let indicator = UIActivityIndicatorView()
+        indicator.hidesWhenStopped = true
+        indicator.frame = CGRect(x: 0, y: 0, width: 50, height: 50)
+        indicator.center = view.center
+        indicator.style = .large
+        indicator.color = .black
+        return indicator
```
공백없는 이 정도 코드 좋은 것 같습니다~ 🔢 
중간에 의미없이 공백이 있으면 읽는 흐름이 끊어지고 무슨 의미로 띄었는지 고민하게 됩니다!

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,183 @@
+    }
+    
+    @objc private func switchCollectionViewLayout() {
+        if productGridCollectionView == nil {
+            setupGridCollectionView()
+        }
```
guard else 문이 아닌 이렇게 if문으로 비교하는 것이 더 명확하네요! 🥇

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,183 @@
+            setupGridCollectionView()
+        }
+        
+        if layoutSegmentedControl.selectedSegmentIndex == 0 {
+            productListCollectionView.isHidden = false
+            productGridCollectionView.isHidden = true
+        } else {
+            productListCollectionView.isHidden = true
+            productGridCollectionView.isHidden = false
+        }
```
여기서도 한번 else 문 없이 같은 동작을 하게 해볼 수 있을까요?

yim2627 replied:
```swift
if layoutSegmentedControl.selectedSegmentIndex == 0 {
    productListCollectionView.isHidden = false
    productGridCollectionView.isHidden = true
    return
}
productListCollectionView.isHidden = true
productGridCollectionView.isHidden = false
```

위 코드 처럼 수정해주었습니다!

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,183 @@
+
+    private func configNavigationBar() {
+        self.navigationItem.titleView = layoutSegmentedControl
+        self.navigationItem.rightBarButtonItem = UIBarButtonItem(image: UIImage(systemName: "plus"), style: .plain, target: self, action: #selector(presentProductRegisterView))
```
저는 이렇게 쭉 한 줄로 이어서 작성하는 것을 선호합니다.
한 줄 띄우고 style 하고 또 한 줄 띄우고 target 쓰고 하는 것은 가독성이 떨어진다고 생각해요.
어디까지 이 괄호가 이어지는 것인지 헷갈리고 그거에 신경써서 읽어야지 안 그러면 한 칸 들여서 새로운 줄이 시작되는 것으로 오해할 수 있습니다.
다만 UIImage 객체를 만드는 부분은 위에 상수에 넣어서 그 상수를 써보는 것이 어떨까요?
그렇게 하면 줄이 더 짧아지고 더 명확해집니다!

ChaminLee replied:
아래와 같이 `plusImage`를 따로 만들어서 넣어주는 방식으로 수정했습니다!
```swift
let plusImage = UIImage(systemName: "plus")
self.navigationItem.rightBarButtonItem = UIBarButtonItem(image: plusImage, style: .plain, target: self, action: #selector(presentProductRegisterView))
```

AppleCEO commented on OpenMarket/OpenMarket/Models/Currency.swift:
```diff
@@ -1,4 +1,15 @@
     case krw = "KRW"
     case usd = "USD"
+    
+    var unit: String {
+        switch self {
+        case .krw:
+            return "KRW"
+        case .usd:
+            return "USD"
+        }
+    }
```
이 unit 은 rawValue 와 같은 값들이 들어있군요.
이렇게 같은 값을 갖는 코드를 추가한 이유가 있을까요?

yim2627 replied:
외부에서 Curreny 타입의 case를 출력할 때 rawValue를 사용하지 않게하려 위처럼 코드를 작성했었습니다.

하지만 rawValue가 이미 존재하고 있으므로 아래와 같은 방법으로 간단하게 수정하려합니다!

```swift
var unit: String {
    return self.rawValue
}
```

AppleCEO commented on OpenMarket/OpenMarket/SceneDelegate.swift:
```diff
@@ -12,10 +12,12 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
-        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
-        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
+        guard let windowScene = (scene as? UIWindowScene) else { return }
+        
```
아래 한 줄을 띄울거면 이 코드를 여러 줄로 작성하는 것은 어떨까요?
저는 이렇게 한 줄로 줄여쓰는 것을 별로 선호하지 않습니다.

ChaminLee replied:
기본적으로 만들어져있던 것을 쓰다보니 인지하지 못했습니다! else문은 개행해서 작성하는 방식으로 수정하겠습니다!

AppleCEO commented on OpenMarket/OpenMarket/Utils/AttributedTextCreator.swift:
```diff
@@ -0,0 +1,48 @@
+            return nil
+        }
+        
+        guard let price = product.price.formattedToDecimal,
+              let bargainPrice = product.bargainPrice.formattedToDecimal else {
+                  return nil
+              }
```
저는 이렇게 let에 맞춰서 줄바꿈을 해주지는 않는 편이에요.
```swift
guard let price = product.price.formattedToDecimal {
  let bargainPrice = product.bargainPrice.formattedToDecimal
  return nil
}
```
위 코드와 착각할 수 있기 때문인데요.
비슷한 실수를 하신 것 같네요.
return 의 위치가 한 칸 들어가있네요.

yim2627 replied:
return과 중괄호를 닫는 부분의 개행이 잘못되어 있음을 알고 아래와 같이 수정해주었습니다.

다만 let들이 불규칙하게 인덴트가 들어가있는 것 보다는 동일선상에 맞추는게 더 가독성이 좋아보일 것이라고 생각하였습니다.
또한  Xcode자체에서 개행을 하면 let의 위치는 동일해지는데, else문의 인덴트가 제대로 잡히지 않음을 알고 이를 주의해야겠다는 생각을 하였습니다! 

```swift
guard let price = product.price.formattedToDecimal,
      let bargainPrice = product.bargainPrice.formattedToDecimal else {
    return nil
}
```

AppleCEO commented on OpenMarket/OpenMarket/Utils/AttributedTextCreator.swift:
```diff
@@ -0,0 +1,48 @@
+        let spacing = " "
+        
+        if product.discountedPrice == 0 {
+            return NSMutableAttributedString
+                .normalStyle(string: "\(product.currency.unit) \(price)")
```
이런 코드도 저는 이해하기 좀 힘드네요.
빠르게 읽다보면 return NSMutableAttributedString 으로 오해할 수 있을 것 같습니다.
짧게 줄여쓰거나 더 멋져보이게 쓸 수 있지만 길더라도 있을 것이라고 예상되는 곳에 코드들이 있는 것이 더 좋은 코드라고 생각합니다~

ChaminLee replied:
넵 한 줄로 작성하도록 수정하였습니다!

AppleCEO commented on OpenMarket/OpenMarket/Utils/AttributedTextCreator.swift:
```diff
@@ -0,0 +1,48 @@
+        priceAttributedText.append(NSMutableAttributedString
+                                    .strikeThroughStyle(string: "\(product.currency.unit) \(price)"))
+        priceAttributedText.append(NSMutableAttributedString.normalStyle(string: spacing))
+        priceAttributedText.append(NSMutableAttributedString
+                                    .normalStyle(string: "\(product.currency.unit) \(bargainPrice)"))
```
이런 코드도 잘못보면
```swift
 priceAttributedText.append(NSMutableAttributedString)
                    .normalStyle(string: "\(product.currency.unit) \(bargainPrice)")
```
이런 코드로 오해할 수 있을 것 같네요!

ChaminLee replied:
위 코멘트 주신 것 처럼, 한 줄로 이어서 작성했습니다. 

고차함수나 체이닝이 길어지는 경우 한 줄에 모두 작성하기가 어려워질 것 같은데, 이런 경우는 개행을 해서 작성을 해도 무방하다고 생각해도 될까요??

AppleCEO replied:
네 그럴 경우에는 괜찮을 것 같습니다. 지금처럼 append 나 매개변수로 값을 보내줄 때는 한 줄로 작성하는게 더 보기 좋을 것 같아요.

AppleCEO reviewed (changes requested):
안녕하세요 도미닉입니다.
@yim2627 & @ChaminLee 고생하셨습니다!
어려웠을텐데 스텝 잘 진행된 것 같습니다.
몇가지 개선하면 좋을 점들 남겨봤습니다~

#궁금한 부분에 대한 답변
1. 상수 관리에 관한 부분 (autolayout의 constant)
- 스탠포드 대학교의 강의를 보더라도 숫자 하나 하나 enum 같은 곳에 모아놓는 것을 권장하긴 합니다. 하드코딩을 지양하는 것이 좋으나 현업에서 빠르게 만들어야하는 경우에는 하드코딩으로 마구 넣는 경우도 있습니다. 우리는 공부하고 있고 더 좋은 구조로 만드는 것을 지금 연습하고 현업에 가서 마구 하드 코딩되어 있어도 그것을 정리하고 더 나은 구조로 리펙토링할 수 있도록 한번 개선해보면 좋겠네요!

2. 보통 safe Area 넘어 view의 상단까지 레이아웃을 주는지 궁금합니다.
- 뷰의 상단까지 레이아웃을 주는 일은 거의 없습니다. 오히려 safe Area 를 제대로 지키지 않은 코드가 현업에 있으면 그것을 safeArea에 맞춰주는 작업을 진행하게 될 수도 있습니다.

3. defaultContentConfiguration()을 사용했을 때 Image의 크기?
보통 default 로 제공되는 cell을 사용한 경우에는 그 크기 그대로 사용했던 것 같습니다. 이미지 크기를 원하는대로 맞춰야한다면 custom 을 하는 것이 좋을 것 같습니다.

4. Deployment Target
- 예전 iOS 버전인 사용자들도 앱을 사용할 수 있게 하기 위해 배포 버전은 쉽게 올리지 않는 편입니다. 그래서 더 최신의 기능을 사용하거나 분기 처리로 최신 버전의 코드와 이전 버전의 코드를 둘 다 작성하는 등 관리하기 힘들기 때문에 iOS 개발자들이 버전을 올리자고 위에 팀장 등에게 건의하는 편입니다. 하지만 아래 버전의 사용자들을 위해 회사는 잘 올리지 않는데요. 그래서 만약 최신 버전에서 달라진 코드들이 있으면 iOS 버전에 따라 분기 처리를 해주기도 합니다.

yim2627 committed:
fix: URLQueryItem 추가 로직 변경 #7

- URLComponents에 URLQueryItem를 append 하는 방식에서 버그가 있었음

ChaminLee committed:
refactor: guard문 if문으로 변경 #7

yim2627 committed:
refactor: 축약형 제거 #7

ChaminLee committed:
refactor: Section -> ProductSection 네이밍 변경 #7

yim2627 committed:
refactor: else문 제거 #7

ChaminLee committed:
refactor: UIImage 생성 코드 분리 #7

yim2627 committed:
refactor: Curreny 내 unit 연산프로퍼티 수정 #7

ChaminLee committed:
style: SceneDelegate내 개행 수정 #7

yim2627 committed:
style: AttributedTextCreator 내 개행 수정 #7

ChaminLee committed:
style: AttributedTextCreator내 chaining 개행 수정 #7

yim2627 committed:
style: 전체적인 코드 개행 수정 #7

ChaminLee committed:
refactor: AttributedTextCreator내 네임스페이스 생성 #7

yim2627 committed:
refactor: MainViewController 네임 스페이스 생성 #7

ChaminLee committed:
refactor: ProductListLayoutCell내 네임스페이스 생성 #7

yim2627 committed:
refactor: ProductGridLayoutCell 네임스페이스 생성 #7

ChaminLee commented:
안녕하세요 도미닉 @AppleCEO 

코드에 코멘트 주신 것 확인하고 개선사항 반영해두었습니다. 이외에 추가적으로 상수관리를 위해 네임 스페이스를 생성해보는 방식을 구현해봤습니다. 

구현하다보니 너무 자잘한 상수들도 관리해주는건가 싶기도 해서, 저희가 생성한 네임 스페이스들이 적합한 정도로 구현된 것인지 궁금합니다!

감사합니다!🙇🏻‍♂️

ChaminLee committed:
fix: Multipart/form-data 형식 수정 #7

- 모델 프로퍼티 오타 수정

AppleCEO commented on OpenMarket/OpenMarket/Models/Currency.swift:
```diff
@@ -1,4 +1,10 @@
 enum Currency: String, Codable {
     case krw = "KRW"
     case usd = "USD"
```
이 rawValue 가 필요할까요?

unit 에 스위치문으로 이 문자열들을 리턴해주는 것과 rawValue 로 선언하는 것 어떤 차이가 있을까요?

yim2627 replied:
raw Value는 디코딩 목적으로 인해 설정해주었습니다. 하지만 실제 코드에선 의미가 명확하지않은 rawValue를 사용하는 것보다 연산프로퍼티를 활용하는 것이 더 좋다고 생각하였습니다.

연산프로퍼티 unit에서 스위치문으로 해당 문자열들을 리턴해주는 경우 호출될때마다 연산이 되고, rawValue로 선언하는 것은 저장프로퍼티(정해진 값)를 가져오는 것의 차이 정도로 알고있습니다... 맞을까요? 🤔
하지만 디코딩 목적으로 생성한 모델이기에 rawValue가 없으면 디코딩이 되지않는 문제가 있어 rawValue로 설정을 해주었습니다.

AppleCEO replied:
아.. 디코딩에서 사용하고 있었군요.
네 그렇다면 연산 프로퍼티를 활용하는 것도 좋은 아이디어네요!

AppleCEO commented on OpenMarket/OpenMarket/Cells/ProductGridLayoutCell.swift:
```diff
@@ -0,0 +1,121 @@
+    }
+    
+    private func configUILayout() {
+        [priceLabel, bargainPriceLabel].forEach {
+            productPriceStackView.addArrangedSubview($0)
+        }
+        
+        [productImageView, productNameLabel, productPriceStackView, productStockLabel].forEach {
+            productStackView.addArrangedSubview($0)
+        }
```
이 부분들도 기본형을 활용해보면 좋겠습니다! 아래 쪽은 for in으로 바꿔보고 어떤 것이 읽기 좋은지 비교해봐도 좋겠네요 ㅎㅎ

ChaminLee replied:
`forEach`에 파라미터를 달아주는 방식으로 수정해보겠습니다! 

`forEach`의 경우, 끊기지 않고 반복문이 돈다는 점에서 커뮤니케이션 관점에서 이점이 있을거라고 생각했습니다! 그래서 위와 같이 짧고 무조건 다 순회하는 반복문의 경우 `forEach`가 더 읽기 좋아보인다고 생각했습니다!

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,199 @@
+}
+
+private enum Design {
+    static let indicatorFrame: CGRect = CGRect(x: 0, y: 0, width: 50, height: 50)
```
여기 50 도 하드코딩을 제거해 볼 수 있을까요?

yim2627 replied:
인디케이터의 프레임 자체를 네임스페이스로 빼주었는데, 네임스페이스로 빼준 프레임의 너비와 높이까지 빼주는게 의미가 크게 있을지 의문이었습니다.

혹시 어떤 이유때문에 50까지 빼주면 좋겠다라고 생각하신지 알 수 있을까요? 네임 스페이스엔 단일 값 (예를 들면 너비, 높이 따로)만 들어가는 것이 더 좋은 방법인지 궁금합니다!

감사합니다! 😀

AppleCEO replied:
단일값만 주로 넣는 것 같아요!
단일 값을 넣고 그 값들을 가져와서 실제 뷰를 그려주는 곳에서 사용하는 식으로 주로 진행합니다.
하드 코딩된 값들을 다 열거형으로 빼보면 어떨까요?

yim2627 replied:
```swift
private enum Design {
   static let indicatorWidth: CGFloat = 50
   static let indicatorHeight: CGFloat = 50
}

indicator.frame = CGRect(origin: .zero, size: CGSize(width: Design.indicatorWidth, height: Design.indicatorHeight))
```

이렇게 수정해주었습니다. 코드가 더 깔끔해졌네요. 
감사합니다!

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,199 @@
+
+    private func configNavigationBar() {
+        self.navigationItem.titleView = layoutSegmentedControl
+        let plusImage = UIImage(systemName: "plus")
+        self.navigationItem.rightBarButtonItem = UIBarButtonItem(image: plusImage, style: .plain, target: self, action: #selector(presentProductRegisterView))
```
이미지를 위에서 상수에 초기화해주는 부분 좋습니다~ 👍

AppleCEO commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,199 @@
+    
+    func createGridLayout() -> UICollectionViewLayout {
+        let itemSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0), heightDimension: .fractionalHeight(1.0))
+        let item = NSCollectionLayoutItem(layoutSize: itemSize)
+        
+        let groupSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0), heightDimension: .absolute(self.view.frame.height * 0.35))
+        let group = NSCollectionLayoutGroup.horizontal(layoutSize: groupSize, subitem: item, count: Design.columnCount)
+        group.interItemSpacing = .fixed(Design.interItemSpacing)
+        
+        let section = NSCollectionLayoutSection(group: group)
+        section.interGroupSpacing = CGFloat(Design.interGroupSpacing)
+        section.contentInsets = Design.sectionEdgeInsets
+        
```
여기 공백들은 어떤 의미가 있을까요?

yim2627 replied:
item, group, section의 레이아웃을 각각 모아주어 보기 쉽게 하려고 공백을 추가해주었습니다.

AppleCEO replied:
그러면 조금씩 다른 동작을 한다고 생각할 수도 있을까요?
한 메소드에서는 하나의 동작만 수행하는 것이 좋습니다!

ChaminLee replied:
말씀주신 것 처럼 기능 분리를 하는 것이 조금 더 좋을 것 같아서 item, group, section을 만드는 메서드를 따로 생성하여 사용해주는 방식으로 개선했습니다!

AppleCEO reviewed (changes requested):
안녕하세요 도미닉입니다
개선하는데 고생 많으셨습니다!
코드들이 많이 깔끔해진 것 같습니다.
네 저는 지금 정도로 네임스페이스 구현한 것 괜찮다고 생각합니다.

몇가지 코멘트 남겨봤습니다.
제가 iPod touch 로 테스트해봤는데 조금 글씨가 깨지는 부분들이 있었습니다.
보너스로 오토레이아웃을 수정해서 iPod touch 에서도 잘 보이도록 개선해볼 수 있을 것 같네요~

ChaminLee committed:
fix: ipod grid layout 대응 #7

- cell의 크기가 줄어들 때 productStackView내 이미지가 줄어들도록 compression, constant 설정

yim2627 committed:
refactor: 고차함수 축약형 수정 #7

ChaminLee commented:
안녕하세요 도미닉 @AppleCEO 

추가미션 주신 오토레이아웃 문제를 개선해봤습니다. 

---

기존의 코드에 이미지의 높이가 고정적으로 설정되어있었던게 문제의 원인이었던 것 같아, 아래와 같이 코드를 추가/수정하여 개선했습니다. 

```swift
// 이미지가 줄어들 수 있도록 제약 추가
productImageView.setContentCompressionResistancePriority(.defaultLow, for: .vertical)
productImageView.heightAnchor.constraint(lessThanOrEqualToConstant: self.contentView.frame.height / 2)
```
![image](https://user-images.githubusercontent.com/45652743/149439656-a78d4a9b-ac9c-44bd-8ff7-eac4a17ae92f.png)

감사합니다!

ChaminLee committed:
feat: imageCache 구현 #7

yim2627 committed:
refactor: 네임스페이스 변경 #7

ChaminLee committed:
refactor: createGridLayout 메서드 기능 분리 #7

- item, group, section을 각각의 메서드에서 만들도록 분리
- gridCell내 stackView들의 distribution 변경

ChaminLee committed:
docs: Step2 readme 작성
